cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DocumentStore.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    public static DocumentStore getInstance(File location, KeyProvider provider) throws DatastoreNotCreatedException {
+        boolean created = checkPathAndCreateIfNeeded(location);
+        try {
+            synchronized (documentStores) {
+                DocumentStore ds = documentStores.get(location);
+                if (ds == null) {
+                    ds = new DocumentStore(location, provider);
+                    documentStores.put(location, ds);
+                    eventBus.post(new DatabaseOpened(ds.databaseName));
+                }
+                if (created) {
+                    eventBus.post(new DatabaseCreated(ds.databaseName));
+                }
+                return ds;
+            }
+        }
+        catch (IOException e) {
+            throw new DatastoreNotCreatedException("Database not found: " + location, e);
+        } catch (SQLException e) {
+            throw new DatastoreNotCreatedException("Database not initialized correctly: " + location, e);
+        } catch (DatastoreException e) {
+            throw new DatastoreNotCreatedException("Datastore not initialized correctly: " + location, e);
+        }
+    }
+
+    private static boolean checkPathAndCreateIfNeeded(File location) {
+        //logger.fine("Datastore path: " + directoryPath);
+
+        boolean created = false;
+        if(!location.exists()){
+            created = location.mkdir();
+        }
+
+        Misc.checkArgument(location.isDirectory(), "Input path is not a valid directory");
+        Misc.checkArgument(location.canWrite(), "Datastore directory is not writable");
+        return created;
+    }
+
+    public void close() {
+        synchronized (documentStores) {
+            DocumentStore ds = documentStores.get(location);
+            if (ds != null) {
+                System.out.println("removing "+location);
+                ((DatabaseImpl)database).close();
+                ((IndexManagerImpl)query).close();
+                documentStores.remove(location);
+            }
+            // TODO warn if not in map?
+        }
+        eventBus.post(new DatabaseClosed(databaseName));
+    }
+
+    public void delete() throws IOException {
+        synchronized (documentStores) {
+            DocumentStore ds = documentStores.remove(location);
+            if (ds != null) {
+                ((DatabaseImpl)database).close();
+                ((IndexManagerImpl)query).close();
+            }
+            if (!location.exists()) {
+                String msg = String.format(
+                        "Datastore %s doesn't exist on disk", location
+                );
+                throw new IOException(msg);
+            } else {
+                FileUtils.deleteDirectory(location);
+                eventBus.post(new DatabaseDeleted(databaseName));
+            }
+        }
+    }
+
+    public static List<File> listInstances() {
+        return new ArrayList<File>(documentStores.keySet());
```

Comment:
I don't think we should return a list of files for this API. We should either return a List of `DocumentStore`s or a change the name.